README.org
#+TITLE: OAuth One

#+BEGIN_HTML
<a href="https://circleci.com/gh/jcf/oauth-one">
  <img src="https://circleci.com/gh/jcf/oauth-one.svg"></img>
</a>
#+END_HTML

* Installation
This project is under active development, and has yet to reach 1.0. As such the
API may change.

#+BEGIN_HTML
  <a href="https://clojars.org/oauth/oauth.one">
    <img src="https://img.shields.io/clojars/v/oauth/oauth.one.svg"></img>
  </a>
#+END_HTML

* Getting started
Require the library with a convenient alias that we can make use of later.

#+begin_src clojure
  (require '[oauth.one :as one])
#+end_src

Create a consumer using the credentials provided by (in this example) Twitter.
The consumer holds on to important URLs, tokens and information about how to
generate OAuth requests. We'll pull our consumer key and secret from environment
variables to avoid adding sensitive credentials to our repository.

#+begin_src clojure
  (def consumer
    (one/make-consumer
     {:access-uri "https://api.twitter.com/oauth/access_token"
      :authorize-uri "https://api.twitter.com/oauth/authorize"
      :callback-uri "http://127.0.0.1:3000/oauth/callback"
      :key (System/getenv "TWITTER_CONSUMER_KEY")
      :request-uri "https://api.twitter.com/oauth/request_token"
      :secret (System/getenv "TWITTER_CONSUMER_SECRET")
      :signature-algo :hmac-sha1}))
#+end_src

** Request token
The ~:callback-uri~ is a local address we can use for testing purposes. All of
the above can and should be pulled from whatever configuration system you're
using your application (I'd recommend a combination of [[https://github.com/weavejester/environ][environ]] and [[https://github.com/plumatic/schema][Schema]]).
Check out my [[https://github.com/jcf/lein-template][lein-template]] for an example of how to pull your configuration from
environment variables and both validate and coerce the data.

Now that you have a consumer you can build a request to request a token.

#+begin_src clojure
  (one/request-token-request consumer)
#+end_src

~request-token-request~ returns a ~clj-http~ compatible hash-map that can be
passed to your favourite HTTP library quite easily. An example request map looks
something like this:

#+begin_src clojure
  {:request-method :post,
   :url "https://api.twitter.com/oauth/request_token",
   :headers
   {"Content-Type" "application/x-www-form-urlencoded",
    "Authorization"
    "OAuth oauth_consumer_key=\"nTCqzIs2jIWSHZw2YNGmw\", oauth_nonce=\"KT-vrp_EqXfYnaCkSartQf3atjj9TK5TxqR44ap25bM\", oauth_signature=\"5Hljpn2TUSeJO4UWR6M8IpxVvuo%3D\", oauth_signature_method=\"HMAC-SHA1\", oauth_timestamp=\"1457741832\", oauth_version=\"1.0\""},
   :throw-exceptions? false}
#+end_src

To send the request simply pass the request map to your favourite HTTP client.

#+begin_src clojure
  (require '[clj-http.client])
  (clj-http.client/request (one/request-token-request consumer))

  (require '[aleph.http])
  (aleph.http/request (one/request-token-request consumer))
#+end_src

** Authorisation URL
The response you get back from the above ~request-token-request~ will include an
~oauth_token~ you can use to ask the user for access to his or her account.

Let's assume you used Aleph to send the ~request-token-request~ request.

#+begin_src clojure
  (def parse-form
    (comp ring.util.codec/form-decode byte-streams/to-string))

  (def request-token-response
    @(manifold.deferred/chain
      (aleph.http/request (one/request-token-request consumer))
      #(update % :body parse-form)))

  ;; (:body request-token-response)
  ;; => {"oauth_token" "t89IVgAAAAAADf8pAAABU2gvEc8",
  ;;     "oauth_token_secret" "XKkTgSNsLNNqRFyZPTf6W4OJT428JeL2",
  ;;     "oauth_callback_confirmed" "true"}

  (authorization-url
   consumer
   {"oauth_token" (get-in request-token-response [:body "oauth_token"])})
#+end_src

~authorization-url~ will return a string URL that you can send a user to in
order to kick of the user-facing part of the OAuth flow.

[[https://dl.dropboxusercontent.com/u/508427/imgs/twitter-oauth-flow-example.png]]

** Access token
When the user surely decides to grant you access to his or her account the OAuth
provider sends the user to your ~:callback-uri~. Some providers allow you to
change this URL, others do not. Last time I check Twitter allowed you to opt in
or out of locking the ~:callback-uri~, which is great in production.

When redirected back you'll receive a couple of ~GET~ parameters containing two
important values needed to finally get your hands on an access token. The URL
will be of the following form:

http://example.com/oauth/callback?{oauth_token,oauth_verifier}

(If you're not familiar with the above notation the curly braces indicate ~GET~
parameters that will show up on the end of your URL.)

You'll need an HTTP endpoint to capture the incoming request from your OAuth
provider, and you'll need to parse the ~oauth_token~ and ~oauth_verifier~. It's
pretty straight forward to do so with ~ring.util.codec~:

#+begin_src clojure
  (defn parse-verifier-url
    [^String url]
    (let [uri (java.net.URI. url)]
      (ring.util.codec/form-decode (.getQuery uri))))

  (parse-verifier-url
   "http://127.0.0.1:3000/oauth/callback?oauth_token=abc123&oauth_verifier=cba321")
  ;; => {"oauth_token" "abc123" "oauth_verifier" "cba321"}
#+end_src

Keep hold of both the ~oauth_token~ and ~oauth_verifier~ because you need them
to get your hands on an access token.

Now we can send a request to get an access token! Hooray!

#+begin_src clojure
  (access-token-request consumer {"oauth_token" "abc123"
                                  "oauth_verifier" "bca321"})
#+end_src

The request will look something like this:

#+begin_src clojure
  {:request-method :post,
   :url "https://api.twitter.com/oauth/access_token",
   :headers
   {"Content-Type" "application/x-www-form-urlencoded",
    "Authorization"
    "OAuth oauth_consumer_key=\"nTCqzIs2jIWSHZw2YNGmw\", oauth_nonce=\"JJpnpVbOpteucb0LfHPMMZk0g2ehQkkFUM8AT3_oj4Q\", oauth_signature=\"e+tgaWSrN5Mzz5yKmNkhkhheQ6U%3D\", oauth_signature_method=\"HMAC-SHA1\", oauth_timestamp=\"1457743849\", oauth_token=\"F096MgAAAAAADf8pAAABU2fcrTM\", oauth_verifier=\"kk9MGzbHcIMnMMJxpecMak7OXvZTCdLo\", oauth_version=\"1.0\""}}
#+end_src

Again, to actually send the request you can use your favourite HTTP library.

The response from this last request will contain the actual ~oauth_token~ and
~oauth_token_secret~. These you'll likely want to store in your database because
they're the credentials you'll use to masquerade as your new user.

** Signing requests
Once you have your hands on both your application credentials, and a user's
token you can send requests on behalf of that user. These requests have to be
cryptographically signed like any other so there's a function provided to make
this easier in your app.

#+begin_src clojure
  (let [request {:request-method :get
                 :url "https://api.twitter.com/account/verify_credentials.json"
                 :query-params {"include_email" "true"
                                "skip_statuses" "true"}}
        access-token {:token "access-token"
                      :secret "access-token-secret"}]
    (one/sign-request consumer request access-token))
#+end_src

You can use the hash-map returned to make a request with your favourite HTTP
client as before.


CHANGELOG.org
#+TITLE: Changes
#+STARTUP: content

* 0.7.0
Upgrade external dependencies to make sure everything's compatible with cutting
edge.

The test suite passed before and after this change. In theory it shouldn't cause
anyone any problems but it's worth double checking for dependency conflicts if
you or another dependency pull in Pandect as they've recently made a fairly
major change to the way they interact with Bouncy Castle.

You should be able to use the previous version numbers of any dependencies
should you not want to upgrade. Just add exclusions to the ~oauth/oauth-one~
artefact in your dependencies.

#+begin_src clojure
  [oauth/oauth-one "0.7.0" :exclusions [org.bouncycastle/bcprov-jdk15on
                                        org.clojure/clojure
                                        pandect
                                        prismatic/schema
                                        ring/ring-codec]]
#+end_src

Make sure to only exclude dependencies you require elsewhere! These dependencies
have to be loaded by something for this library to work.

* 0.6.0
*No breaking changes.*

You can now sign request more easily. Previously, you'd have to create OAuth
headers yourself, and associate in the OAuth token from an access token. Then
you'd need to pass in your access token secret for the signing process. The
whole dance looked something like this:

#+begin_src clojure
  (one/sign-request (one/make-consumer config)
                    {:oauth-headers (merge (one/make-oauth-headers consumer)
                                           {"oauth_token" token})
                     :request-method :get
                     :url "https://www.example.com/api"}
                    secret)
#+end_src

Based on feedback from @ilevd it was apparent this is a little clunky, and
exposing the OAuth protocol wasn't very user friendly so I've extended
~sign-request~ to support both the old version and this easier API:

#+begin_src clojure
  (one/sign-request (one/make-consumer config)
                    {:request-method :get
                     :url "https://www.example.com/api"}
                    {:token "access-token"
                     :secret "access-token-secret"})
#+end_src

Notice that you can pass an access token map as the third argument to
~sign-request~, and internally we juggle things around to do the right thing.

And, you can still pass in OAuth headers inside the request map if you so wish.

* 0.5.0
Fix escaping of characters by sticking more closely to [[https://www.ietf.org/rfc/rfc3986.txt][RFC 3986]].

This release also removes the dependency on crypto-random, as we now generate
nonces more directly.

* 0.4.0
Avoid non-word characters in nonces.

#+BEGIN_QUOTE
The oauth_nonce parameter is a unique token your application should
generate for each unique request. Twitter will use this value to
determine whether a request has been submitted multiple times. The
value for this request was generated by base64 encoding 32 bytes of
random data, and stripping out all non-word characters, but any
approach which produces a relatively random alphanumeric string should
be OK here.
#+END_QUOTE

https://twittercommunity.com/t/how-to-generate-an-oauth-nonce/1307

* 0.3.0
** Better support for query parameters
Previously, only form parameters were supported as that's all I was using. Now
that query parameters are required, you can pass params on the end of the URL,
or in a separate map of query params.

#+begin_src clojure
  (one/sign-request
   consumer
   {:url "https://example.com/?foo=bar"
    :query-params {"baz" "qux"}})
#+end_src

Both the ~foo~ parameter and ~baz~ parameters above will be merged into a single
map of signed parameters as per the OAuth 1.0 spec.

* 0.2.0
** Add ~parse-auth-header~ utility function
Add ~parse-auth-header~ function that was previously only used in the test
suite. This function is really handy in other test suites too!

#+begin_src clojure
  (require '[oauth.one :as one])

  (one/parse-auth-header "OAuth oauth_consumer_key=\"key\"")
  ;; => {"oauth_consumer_key" "key"}
#+end_src

The function works both with and without the ~OAuth~ prefix.

** Improved signed request
There are some fairly significant *breaking* changes in this release for anyone
who made direct use of ~signed-request~ and its associated schema.

~signed-request~ has been renamed to ~sign-request~ and has slightly modified
behaviour. Firstly, you can now omit OAuth headers when calling ~signed-request~
if you're trying to send an authenticated request.

#+begin_src clojure
  (require '[oauth.one :as one])

  (def consumer
    (one/make-consumer {:key "etc"}))

  (one/sign-request
   consumer
   {:request-method :get
    :url "https://api.twitter.com/account/verify_credentials"})
#+end_src

You can create a header map directly via the new, public ~make-oauth-headers~
function if you want to associate some state in before creating a signed request
via ~sign-request~.

#+begin_src clojure
  (one/sign-request
   consumer
   {:request-method :get
    :oauth-headers
    (assoc
     (one/make-oauth-headers consumer)
     "oauth_extension" "etc")
    :url "https://api.twitter.com/account/verify_credentials"})
#+end_src

You may notice, ~:oauth-headers~ used to be called ~:oauth-params~. The tests
have been updated accordingly, and are a good place to look for up-to-date
examples of how to use the library.

In addition, the version of Schema being used has been upgraded to 1.1.0.


src/oauth/one.clj
(ns oauth.one
  "Provides functions for generating requests, and parsing responses necessary
  for integrating with three-legged OAuth 1.0 providers like Twitter.

  All Token requests and Protected Resources requests MUST be signed by the
  Consumer and verified by the Service Provider. The purpose of signing requests
  is to prevent unauthorized parties from using the Consumer Key and Tokens when
  making Token requests or Protected Resources requests. The signature process
  encodes the Consumer Secret and Token Secret into a verifiable value which is
  included with the request.

  OAuth does not mandate a particular signature method, as each implementation
  can have its own unique requirements. The protocol defines three signature
  methods: HMAC-SHA1, RSA-SHA1, and PLAINTEXT, but Service Providers are free to
  implement and document their own methods. Recommending any particular method
  is beyond the scope of this specification.

  The Consumer declares a signature method in the `oauth_signature_method`
  parameter, generates a signature, and stores it in the `oauth_signature`
  parameter. The Service Provider verifies the signature as specified in each
  method. When verifying a Consumer signature, the Service Provider SHOULD check
  the request nonce to ensure it has not been used in a previous Consumer
  request.

  The signature process MUST NOT change the request parameter names or values,
  with the exception of the `oauth_signature` parameter.

  For more information refer to the OAuth 1.0 specification at
  http://oauth.net/core/1.0/#signing_process."
  (:require [clojure.string :as str]
            [pandect.core :as pandect]
            [ring.util.codec :as codec]
            [schema.core :as s]))

;; -----------------------------------------------------------------------------
;; Schema

(def ^:private signature-algos
  "Mapping from keyword algorithm used in a consumer to the string version used
  in the OAuth Authorization header."
  {:hmac-sha1 "HMAC-SHA1"
   :plaintext "PLAINTEXT"
   :rsa-sha1 "RSA-SHA1"})

(def SignatureAlgo
  (apply s/enum (keys signature-algos)))

(def SignatureMethod
  (apply s/enum (vals signature-algos)))

(def ConsumerConfig
  {:access-uri s/Str
   :authorize-uri s/Str
   :callback-uri s/Str
   :key s/Str
   :request-uri s/Str
   :secret s/Str
   :signature-algo SignatureAlgo})

(def ^:private Map
  "Hash-map of keyword or string to any value"
  {(s/either s/Keyword s/Str) s/Any})

(def OAuthAuthorization
  "Valid attributes and corresponding values allowed in the unsigned OAuth
  Authorization header."
  {(s/optional-key "oauth_callback") s/Str
   (s/optional-key "oauth_token") s/Str
   (s/optional-key "oauth_verifier") s/Str
   (s/optional-key "oauth_version") s/Str
   (s/required-key "oauth_consumer_key") s/Str
   (s/required-key "oauth_nonce") s/Str
   (s/required-key "oauth_signature_method") SignatureMethod
   (s/required-key "oauth_timestamp") (s/either s/Str s/Int)})

(def SignedOAuthAuthorization
  "Signed version of the valid attributes and corresponding values allowed in
  the OAuth Authorization header. See `OAuthAuthorization`."
  (assoc OAuthAuthorization (s/required-key "oauth_signature") s/Str))

(def ^:private RequestTokenParams
  {(s/optional-key "oauth_callback") s/Str})

(def ^:private AuthorizationParams
  (assoc Map (s/optional-key :callback-uri) s/Str
         (s/optional-key "oauth_token") s/Str))

(def ^:private RequestMethod
  "Valid HTTP request methods"
  (s/enum :delete :get :head :patch :post :put :trace))

(def ^:private Request
  "A clj-http compatible request map that is also OAuth 1.0 compatible."
  {(s/optional-key :form-params) Map
   (s/optional-key :headers) {s/Str s/Str}
   (s/optional-key :query-params) Map
   :request-method RequestMethod
   :url s/Str})

(def ^:private OAuthRequest
  "clj-http compatible request map with required `:oauth-headers` that will be
  used to generate a signature."
  (assoc Request (s/optional-key :oauth-headers) OAuthAuthorization))

(def ^:private SignedRequest
  "Signed OAuth request with an Authorization header."
  (assoc Request
         (s/optional-key :headers)
         {(s/required-key "Authorization") s/Str
          s/Str s/Str}))

(def ^:private AccessToken
  {(s/optional-key :token) (s/maybe s/Str)
   (s/optional-key :secret) (s/maybe s/Str)})

(def ^:private AccessTokenArg
  (s/conditional
   map? AccessToken
   :else s/Str))

;; -----------------------------------------------------------------------------
;; Utils

(defn- filter-vals
  [m]
  (into (empty m) (filter val m)))

(s/defn ^:private split-url :- [(s/one s/Str "base-url") (s/maybe {s/Str s/Str})]
  [url]
  (let [uri (java.net.URI. url)]
    [(str (.getScheme uri) "://" (.getAuthority uri) (.getPath uri))
     (some-> uri .getQuery codec/form-decode)]))

;; -----------------------------------------------------------------------------
;; OAuth encoded

(s/defn oauth-encode :- s/Str
  "Encode string according to RFC 3986."
  [s]
  (-> (str s)
      (java.net.URLEncoder/encode "UTF-8")
      (.replace "+" "%20")
      (.replace "*" "%2A")
      (.replace "%7E" "~")))

(s/defn oauth-decode :- s/Str
  "Decode an RFC 3986 encoded string."
  [^String s]
  (java.net.URLDecoder/decode s "UTF-8"))

;; -----------------------------------------------------------------------------
;; Nonce

(def ^:private ^java.util.Random secure-random
  (java.security.SecureRandom/getInstance "SHA1PRNG"))

(def ^:private nonce-chars
  (mapv #(Character/toString (char %))
        (concat (range 48 57) (range 65 90) (range 97 122))))

(s/defn nonce :- s/Str
  [length :- s/Int]
  (let [n (count nonce-chars)]
    (str/join (repeatedly length #(nth nonce-chars (.nextInt secure-random n))))))

;; -----------------------------------------------------------------------------
;; Consumer

(defrecord Consumer
    [access-uri authorize-uri callback-uri key secret signature-algo])

(s/defn make-consumer :- Consumer
  "Create a new consumer instance with necessary URIs, key and secret."
  [config :- ConsumerConfig]
  (map->Consumer config))

;; -----------------------------------------------------------------------------
;; Request signing

(s/defn parse-auth-header :- {s/Str s/Str}
  "The inverse of `auth-headers->str`."
  [s :- s/Str]
  (reduce
   #(if-let [[_ k v] (re-find #"(.*?)=\"(.*?)\"" %2)]
      (assoc %1 k (codec/url-decode v))
      %1)
   {}
   (str/split (str/replace s #"(?i)^oauth\s+" "") #",\s+")))

(s/defn auth-headers->str :- s/Str
  "The OAuth Protocol Parameters are sent in the Authorization header the
  following way:

  1. Parameter names and values are encoded per Parameter Encoding.
  2. For each parameter, the name is immediately followed by an ‘=’ character
     (ASCII code 61), a ‘”’ character (ASCII code 34), the parameter value (MAY
     be empty), and another ‘”’ character (ASCII code 34).
  3. Parameters are separated by a comma character (ASCII code 44) and OPTIONAL
     linear whitespace per [RFC2617](http://oauth.net/core/1.0/#RFC2617).
  4. The OPTIONAL realm parameter is added and interpreted per
     [RFC2617](http://oauth.net/core/1.0/#RFC2617), section 1.2.

  http://oauth.net/core/1.0/#auth_header"
  [m :- {s/Str s/Any}]
  (->> m
       (map #(format "%s=\"%s\""
                     (key %)
                     (-> % val oauth-encode)))
       (str/join ", ")))

(s/defn ->seconds :- s/Int
  [millis :- s/Int]
  (quot ^int millis 1000))

(s/defn sign :- s/Str
  [consumer :- Consumer oauth-token-secret :- (s/maybe s/Str) data :- s/Str]
  (let [{:keys [secret signature-algo]} consumer]
    (case signature-algo
      :hmac-sha1
      (codec/base64-encode
       (pandect/sha1-hmac-bytes
        data
        (format "%s&%s"
                (oauth-encode secret)
                (oauth-encode (or oauth-token-secret ""))))))))

(s/defn ^:private base-string :- s/Str
  "http://oauth.net/core/1.0/#anchor14

  The Signature Base String is a consistent reproducible concatenation of the
  request elements into a single string. The string is used as an input in
  hashing or signing algorithms. The HMAC-SHA1 signature method provides both a
  standard and an example of using the Signature Base String with a signing
  algorithm to generate signatures. All the request parameters MUST be encoded
  as described in Parameter Encoding prior to constructing the Signature Base
  String.

  The following items MUST be concatenated in order into a single string. Each
  item is encoded and separated by an ‘&’ character (ASCII code 38), even if
  empty.

  1. The HTTP request method used to send the request. Value MUST be uppercase,
     for example: `HEAD`, `GET`, `POST`, etc.
  2. The request URL from Section 9.1.2.
  3. The normalized request parameters string from Section 9.1.1."
  [method :- (s/either s/Keyword s/Str)
   uri :- s/Str
   params :- Map]
  {:pre [(sorted? params)]}
  (format "%s&%s&%s"
          (-> method name str/upper-case)
          (oauth-encode uri)
          (->> params
               filter-vals
               (map (fn [[k v]]
                      (format "%s=%s" k (oauth-encode (str v)))))
               (str/join "&")
               oauth-encode)))

(s/defn make-oauth-headers :- OAuthAuthorization
  [consumer :- Consumer]
  (sorted-map
   "oauth_consumer_key"     (:key consumer)
   "oauth_nonce"            (nonce 32)
   "oauth_signature_method" (-> consumer :signature-algo signature-algos)
   "oauth_timestamp"        (->seconds (System/currentTimeMillis))
   "oauth_version"          "1.0"))

(s/defn ^:private sign-request-impl :- SignedRequest
  ([consumer oauth-request] (sign-request-impl consumer oauth-request {}))
  ([consumer :- Consumer
    oauth-request :- OAuthRequest
    access-token :- AccessToken]
   (let [{:keys [form-params
                 oauth-headers
                 query-params
                 request-method
                 url]} oauth-request
         {access-token-secret :secret} access-token
         [base-url url-query-params] (split-url url)
         params (merge oauth-headers url-query-params query-params form-params)
         base-string (base-string request-method base-url params)
         signed-params (assoc oauth-headers
                              "oauth_signature"
                              (sign consumer access-token-secret base-string))]
     (with-meta
       (-> oauth-request
           (dissoc :oauth-headers)
           (assoc-in [:headers "Content-Type"]
                     "application/x-www-form-urlencoded")
           (assoc-in [:headers "Authorization"]
                     (str "OAuth " (auth-headers->str signed-params))))
       {:base-url base-url
        :query-params query-params
        :base-string base-string
        :signed-params signed-params}))))

(s/defn ^:private normalize-access-token-arg :- AccessToken
  [x :- (s/maybe AccessTokenArg)]
  (cond
    (and (map? x) (contains? x :token) (contains? x :secret)) x
    (string? x) {:secret x}
    (nil? x) {}
    :else (throw (IllegalArgumentException.
                  (str "access-token passed to `sign-request` must be a map of "
                       "`:token` and `:secret`, or "
                       "an access token secret as a string.")))))

(s/defn sign-request :- SignedRequest
  ([consumer oauth-request]
   (sign-request consumer oauth-request nil))
  ([consumer :- Consumer
    {:keys [oauth-headers] :as oauth-request} :- OAuthRequest
    access-token-arg :- (s/maybe AccessTokenArg)]
   (let [;; In order to support both the string and map version of the
         ;; `access-token` we need to normalise the input into the preferred map
         ;; form.
         ;;
         ;; If the token is something other than a valid map or string we throw
         ;; an `IllegalArgumentException`.
         {:keys [token]
          :as access-token} (normalize-access-token-arg access-token-arg)
         ;; To allow optional `oauth-headers` from users calling this function
         ;; directly we conditionally use a new map of sorted OAuth headers.
         ;;
         ;; When we've got an OAuth token to sign the request with, we add that
         ;; to the OAuth headers for use in `sign-request-impl`.
         oauth-headers (cond-> (or oauth-headers (make-oauth-headers consumer))
                         (some? token) (assoc "oauth_token" token))]
     (sign-request-impl consumer
                        (assoc oauth-request :oauth-headers oauth-headers)
                        access-token))))

;; -----------------------------------------------------------------------------
;; Request token

(s/defn request-token-request
  "Generate a clj-http compatible request map that will request a token from the
  provider associated with `consumer`.

  http://oauth.net/core/1.0/#auth_step1

  The Consumer obtains an unauthorized Request Token by asking the Service
  Provider to issue a Token. The Request Token’s sole purpose is to receive
  User approval and can only be used to obtain an Access Token.

  To obtain a Request Token, the Consumer sends an HTTP request to the Service
  Provider’s Request Token URL. The Service Provider documentation specifies
  the HTTP method for this request, and HTTP POST is RECOMMENDED.

  Note, if you override the \"oauth_callback\" via `params`, you need to pass
  the same callback URI to `authorization-url`."
  ([consumer :- Consumer] (request-token-request consumer {}))
  ([consumer :- Consumer params :- RequestTokenParams]
   (sign-request-impl
    consumer
    {:oauth-headers
     (assoc
      (make-oauth-headers consumer)
      "oauth_callback" (get params "oauth_callback" (:callback-uri consumer)))
     :request-method :post
     :url (:request-uri consumer)})))

;; -----------------------------------------------------------------------------
;; User authorisation

(s/defn authorization-url
  "Generate a provider-specific authorisation URL that you send the user's agent
  (aka. browser) to typically via an HTTP redirect.

  Optional `params` can be passed to append to the authorisation URL via a query
  string.

  `params` may contain an \"oauth_callback\" to override any callback URI in the
  consumer. This can be useful when you need to pass some state for CSRF
  protection to the OAuth provider.

  Note, if you override the \"oauth_callback\" via `params`, you need to pass
  the same callback URI to `request-token-request`."
  ([consumer :- Consumer] (authorization-url consumer {}))
  ([consumer :- Consumer params :- AuthorizationParams]
   (format
    "%s?%s"
    (:authorize-uri consumer)
    (-> {"oauth_callback" (:callback-uri consumer)}
        (merge params)
        filter-vals
        codec/form-encode))))

;; -----------------------------------------------------------------------------
;; Access token request

(s/defn ^:always-validate access-token-request
  "Generate a signed request that will ask the OAuth provider for an access
  token.

  This request must contain the token and verifier provided by the OAuth
  provider when they redirect back to your application after you send someone to
  the authorisation URL generated by `authorization-url`."
  [consumer :- Consumer
   creds :- {(s/optional-key "oauth_token") s/Str
             (s/optional-key "oauth_verifier") s/Str}]
  (sign-request-impl
   consumer
   {:oauth-headers
    (merge
     (make-oauth-headers consumer)
     (when-let [token (get creds "oauth_token")]
       {"oauth_token" token})
     (when-let [verifier (get creds "oauth_verifier")]
       {"oauth_verifier" verifier}))
    :request-method :post
    :url (:access-uri consumer)}))
